hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/DirectoryDeletingService.java

```diff
@@ -169,27 +210,27 @@ public BackgroundTaskResult call() {
             = new ArrayList<>((int) remainNum);
 
         Table.KeyValue<String, OmKeyInfo> pendingDeletedDirInfo;
-
-        try (TableIterator<String, ? extends KeyValue<String, OmKeyInfo>>
-                 deleteTableIterator = getOzoneManager().getMetadataManager().
-            getDeletedDirTable().iterator()) {
-          // This is to avoid race condition b/w purge request and snapshot chain updation. For AOS taking the global
-          // snapshotId since AOS could process multiple buckets in one iteration.
+        // This is to avoid race condition b/w purge request and snapshot chain updation. For AOS taking the global
+        // snapshotId since AOS could process multiple buckets in one iteration.
+        try {
           UUID expectedPreviousSnapshotId =
-              ((OmMetadataManagerImpl)getOzoneManager().getMetadataManager()).getSnapshotChainManager()
+              ((OmMetadataManagerImpl) getOzoneManager().getMetadataManager()).getSnapshotChainManager()
                   .getLatestGlobalSnapshotId();
 
           long startTime = Time.monotonicNow();
-          while (remainNum > 0 && deleteTableIterator.hasNext()) {
-            pendingDeletedDirInfo = deleteTableIterator.next();
+          while (remainNum > 0) {
+            pendingDeletedDirInfo = deletedDirSupplier.get();
```

Comment:
Can't we get multiple pending directories to delete here, so that to decrease probability of all threads to wait for `synchronized get` at the same time.

Reply:
IMO,
get() will be synchronized for 1 unit of time, and processing will take 5-10 unit of time minimum as, it processing involves,
- search in file table
- search in directory table
- prepare request
So this contention will not have much impact in processing with default 10 threads. Just keep simple instead of having cache / or other mechanism.

Reply:
I think we can micro-benchmark this to test and be sure which is the best approach. If there's no speedup to retrieving batches of 10,100, etc per get then +1 for keeping it simple with a single get, but I think something like this we can test out to get a conclusive answer.

Reply:
There is a concern of getting 10K batches, mostly it will be wasted as per logic,
- Get parent
-- scan subdir and subfile
-- if subdir/subfile itself meet the limit, then it will not read remaining 9.999k records in this iteration.

Limit: ratis size OR limit being configured, and this limit is per thread.

So its suggested not to have cache. Additionally, multiple iterator have concern of duplicate reading or managing gaps between threads.

So We can test, various scenario if iteration can be completed in a minute.
1) 10 threads, each thread limit: 10k, flat director only within a parent and count: 200k
2) 50 threads, each limit: 10k, flat directory only within a parent and count: 2M
3) 10 thread, each limit 50k, flat directory only within a parent and count: 2M

This can check comparision of 2 and 3 case, with info log for time taken as already log present.
cc: @aryangupta1998